src/Sentry.OpenTelemetry/SentrySpanProcessor.cs

```diff
@@ -196,6 +207,44 @@ public override void OnEnd(Activity data)
         span.Finish(status);
 
         _map.TryRemove(data.SpanId, out _);
+
+        // Housekeeping
+        PruneFilteredSpans();
+    }
+
+    /// <summary>
+    /// Clean up items that may have been filtered out.
+    /// See https://github.com/getsentry/sentry-dotnet/pull/3198
+    /// </summary>
+    internal void PruneFilteredSpans(bool force = false)
+    {
+        if (!force && !NeedsPruning())
+        {
+            return;
+        }
+
+        foreach (var mappedItem in _map)
+        {
+            var (spanId, span) = mappedItem;
+            var activity = span.GetFused<Activity>();
+            if (activity is { Recorded: false, IsAllDataRequested: false })
+            {
+                _map.TryRemove(spanId, out _);
+            }
+        }
+    }
+
+    private bool NeedsPruning()
+    {
+        lock (_pruningLock)
```

Comment:
This reads as potential contention code and since we're just doing some path with the date is there a way we can do something atomic here instead with Interlocked?

Reply:
suggestion: unix timestamp as seconds and interlocked.compareExchange

Reply:
I've replaced lock with some rough equivalent using Interlocked... In the Benchmarks I ran, I wasn't able to detect a significant performance difference between the two, but that might have been a shortcoming in the Benchmark itself. I'll have another crack at it as intuitively the two should behave quite differently.

Reply:
OK, I put together some new benchmarks... interestingly, `lock` actually comes out a bit faster than either `Interlocked` or `SemaphoreSlim`. The difference isn't wild and I'm not sure how much I trust these - I'm worried there's some error in my design of the benchmark tests themselves. [Code here](https://github.com/jamescrosswell/LockBenchmark/blob/main/BenchmarkDotNet.Artifacts/results/LockBenchmark.LockBenchmarks-report-github.md) however.

Reply:
> interestingly, lock actually comes out a bit faster than either Interlocked or SemaphoreSlim. 

I really don't trust the benchmark then :)

Reply:
the link 404's to me

Reply:
Whoops... that repo was private. Just made it public.